File: benchmarks/extreme_benchmark.go
package main

import "fmt"

func fib(n int64) int64 {
	if n <= 1 {
		return n
	}
	return fib(n-1) + fib(n-2)
}

func bench_fib() int64 {
	return fib(42)
}

func bench_alloc() int64 {
	count := int64(10000000)
	var sum int64 = 0
	for i := int64(0); i < count; i++ {
		// Go GC will run eventually, but we act dumb
		ptr := new(int64)
		*ptr = i
		sum += *ptr
	}
	return sum
}

func process_request(id int64) int64 {
	var acc int64 = 0
	for i := int64(0); i < 1000; i++ {
		acc += i * id
	}
	return acc
}

func bench_requests() int64 {
	// Sequential to match others fairness
	total := int64(1000000)
	var completed int64 = 0
	for i := int64(0); i < total; i++ {
		if process_request(i) > 0 {
			completed++
		}
	}
	return completed
}

func bench_sieve() int64 {
	limit := 50000000
	isPrime := make([]bool, limit)
	for i := range isPrime {
		isPrime[i] = true
	}

	for i := 2; i*i < limit; i++ {
		if isPrime[i] {
			for j := i * i; j < limit; j += i {
				isPrime[j] = false
			}
		}
	}

	var count int64 = 0
	for i := 2; i < limit; i++ {
		if isPrime[i] {
			count++
		}
	}
	return count
}

func main() {
	f := bench_fib()
	a := bench_alloc()
	r := bench_requests()
	s := bench_sieve()
	fmt.Printf("%d\n", f+a+r+s)
}

File: benchmarks/extreme_benchmark.rs
fn fib(n: i64) -> i64 {
    if n <= 1 { return n; }
    fib(n - 1) + fib(n - 2)
}

fn bench_fib() -> i64 {
    fib(42)
}

fn bench_alloc() -> i64 {
    let count = 10000000;
    let mut sum = 0;
    for i in 0..count {
        let b = Box::new(i);
        sum += *b;
        std::mem::forget(b); // No free, leak like Aether
    }
    sum
}

fn process_request(id: i64) -> i64 {
    let mut acc = 0;
    for i in 0..1000 {
        acc += i * id;
    }
    acc
}

fn bench_requests() -> i64 {
    let total = 1000000;
    let mut completed = 0;
    for i in 0..total {
        if process_request(i) > 0 {
            completed += 1;
        }
    }
    completed
}

fn bench_sieve() -> i64 {
    let limit = 50000000;
    let mut is_prime = vec![true; limit];
    
    let mut i = 2;
    while i * i < limit {
        if is_prime[i] {
            let mut j = i * i;
            while j < limit {
                is_prime[j] = false;
                j += i;
            }
        }
        i += 1;
    }
    
    let mut count = 0;
    for i in 2..limit {
        if is_prime[i] { count += 1; }
    }
    count
}

fn main() {
    let f = bench_fib();
    let a = bench_alloc();
    let r = bench_requests();
    let s = bench_sieve();
    // inhibit optimization
    println!("{}", f + a + r + s); 
}
